subsys/mgmt/mcumgr/lib/mgmt/src/mgmt.c

```diff
@@ -65,30 +70,54 @@ mgmt_register_group(struct mgmt_group *group)
 	sys_slist_append(&mgmt_group_list, &group->node);
 }
 
-void
-mgmt_ntoh_hdr(struct mgmt_hdr *hdr)
+#if defined(CONFIG_MGMT_NOTIFICATION_HOOKS)
+void mgmt_register_callback(struct mgmt_callback *callback)
 {
-	hdr->nh_len = sys_be16_to_cpu(hdr->nh_len);
-	hdr->nh_group = sys_be16_to_cpu(hdr->nh_group);
+	sys_slist_append(&mgmt_callback_list, &callback->node);
 }
 
-void
-mgmt_hton_hdr(struct mgmt_hdr *hdr)
+void mgmt_unregister_callback(struct mgmt_callback *callback)
 {
-	hdr->nh_len = sys_cpu_to_be16(hdr->nh_len);
-	hdr->nh_group = sys_cpu_to_be16(hdr->nh_group);
+	(void)sys_slist_find_and_remove(&mgmt_callback_list, &callback->node);
+}
+
+bool mgmt_notify_callbacks(uint32_t event, uint16_t group, uint8_t id, void *data,
```

Comment:
I think that at least 3 first parameters should be part of single "context" object. These things will be the same through processing of single group, loading them separately each time you want to send notification will be overkill.
Instead of returning 'bool' status we should define some specific error set that will be returned and then translated in lower level.

Reply:
I was keeping them like this to aid migration of the existing fs_mgmt API. In terms of returning a status code, my idea for that was to use the provided struct, for messages that support it, because the vast majority of callbacks have no status on influence on running other callbacks, at present it's just the fs_mgmt file read/write callback and img_mgmt chunk callback that do checks on the return code

Reply:
> I was keeping them like this to aid migration of the existing fs_mgmt API. 

You have already announced rework, so you are free to rework it completely.
We are loosing compatibility anyway, so why bother?
Instead of using the "coordinates" to identify callback point you may just go for two:
1) Group/mcumgr subsystem
2) Notification point

These can be encoded into single 32 bit event code easily.

For example (naming and numbers may differ)
```
// 8 codes reserved main areas mgmt (group registration, etc), SMP protocol, buffer allocations
#define MCUMGR_MGMT_NOTIFY_ID(area, point) (area << 8 | point)
#define MCUMGR_MGMT_NOTIFY_AREA_MGMT 1
#define MCUMGR_MGMT_NOTIFY_MGMT_REGISTER_GRP MCUMGR_MGMT_NOTIFY_ID(MCUMGR_MGMT_NOTIFY_AREA_MGMT, 10)
....
// 16 codes, starting at 16 reserved for transports. Will there be more?
#define MCUMGR_MGMT_NOTIFY_AREA_TRANSPORT 9
....
```

Then it is enough to do:
```
if (point == MCUMGR_MGMT_NOTIFY_ID(MCUMGR_MGMT_...AREA..., MCUMGR_MGMT_....))
```
or
```
switch (id)
{
case MCUMGR_MGMT_NOTIFY_ID(MCUMGR_MGMT_...AREA..., MCUMGR_MGMT_....):
...
```

To process events. 

You have chance to design single callback infrastructure to serve  informative/review/security purposes? Why not take time to do so?

> In terms of returning a status code, my idea for that was to use the provided struct, for messages that support it, because the vast majority of callbacks have no status on influence on running other callbacks (...)

Things returned by callbacks have to be clearly defined so that user would know what to return and how does this affect mcumgr code, this would require defining "callback points" to identify and pair them with values. MCUmgr code can only react to what it knows and we do not want everyone to expand logic in MCUmgr, to serve notifications.

For example (naming is just for example):
event point `MCUMGR_MGMT_NOTIFY_GRP_FS_DOWNLOAD_START` expects one of:
```
 MCUMGR_MGMT_NOTIFY_FS_BUSY -- can not handle processing now, file will not be downloaded, proper BUSY response will be reported to client;
 MCUMGR_MGMT_NOTIFY_FS_OK -- will process download;
 MCUMGR_MGMT_NOTIFY_FS_UNAVAILABLE -- pretend file is not there;
 MCUMGR_MGMT_NOTIFY_FS_ACCESS_DENIED -- not authorized or missing authorization token;
 MCUMGR_MGMT_NOTIFY_FS_PM_DEVICE_DOWN -- power management failed to bring device up, used when aggressive power management is used and notification functions are used to power up/down devices at various points of download.
All other responses, for security, are treated as MCUMGR_MGMT_NOTIFY_FS_ACCESS_DENIED.
```

This really needs thinking over what we expect of it and reworking.

Reply:
> These can be encoded into single 32 bit event code easily.

That system is very limiting and not usable, it assumes a maximum of 8 mgmt groups, there are users of mcumgr that already have in excess of 20 groups in a single product, so 8 is not sufficient.

> Things returned by callbacks have to be clearly defined so that user would know what to return and how does this affect mcumgr code, this would require defining "callback points" to identify and pair them with values. MCUmgr code can only react to what it knows and we do not want everyone to expand logic in MCUmgr, to serve notifications.

Not a fan of this at all, this then introduces new return codes specific for each event. If a status is going to be returned, it would need to be one of `MGMT_ERR_*` (which needs new error codes adding)

Reply:
> > These can be encoded into single 32 bit event code easily.
> 
> That system is very limiting and not usable, it assumes a maximum of 8 mgmt groups, there are users of mcumgr that already have in excess of 20 groups in a single product, so 8 is not sufficient.

It is an example, you can take 24 bits to encode groups/transport/mgmt subsystem and 8 bit to encode event points, even though this should be overkill, because it allows to put points in 255 lines.

> 
> Not a fan of this at all, this then introduces new return codes specific for each event. If a status is going to be returned, it would need to be one of `MGMT_ERR_*` (which needs new error codes adding)

Then design common code set like `ERRNO` does, and let users extend them for the purpose of their groups.

Reply:
8 bits means a maximum of 8 callbacks for a whole mgmt group, that is not sufficient. The mgmt group ID itself is 16 bits, so that field cannot be any less than 16 bits for that alone (will rework it to use the group ID directly rather than a bitmask for that part), can use a 32 bit field for it and designate the first few bits to determine the type. Stuffing all fields together makes no sense anyway, it's just going to use more flash to stuff the bits and then more flash for every callback function where it unstuffs the bits which can instead be provided as 2 direct variables instead.

Reply:
> 8 bits means a maximum of 8 callbacks for a whole mgmt group, that is not sufficient. 

8 bits, means 255 identifiers, not counting 0, for callback points. Then you have 24 bits where you can use 8 bit to identify what additional 16 bits mean: groups, transports, etc.

Why would you use bit flags flags? It is not convenient to mark where did notification come from with flag. You can exactly identify path by giving it direct identifier.

> Stuffing all fields together makes no sense anyway, it's just going to use more flash to stuff the bits and then more flash for every callback function where it unstuffs the bits which can instead be provided as 2 direct variables instead.

I am not sure about this, note that calling function requires putting all the parameters in registers anyway, which is ops that go to flash, and storing some of them on stack which is ops in flash and space on stack.

Reply:
> It's very easy from a code perspective to enable/disable/check which bits are enabled. Not using bitmask values means having to then have individual lists for every event to register for, if a function wants to register for 1 fs_mgmt event and 3 img_mgmt events, it now needs to register 4 times separately, and the callback system needs to have n (n=events, where in this case n=4) sets of handler details to check through instead of (n=handlers, where in this case n=1, or n=2 if the bitmasks for each group are different). In a system with 4 callbacks each checking 1 event from a single group, with a bitmask value, n=4, with separate events, n=4 In a system with 4 callbacks each checking 3 events from a single group, with a bitmask value, n=4, with separate events, n=12 With 32 or 64-bit size, in the current system (assuming it is re-written so that the bitmask values are reused for each mgmt group) that allows for 32 or 64 callbacks per group with a linear access time and flat variable size, switching to incrementing values instead would allow for 255 events per group with an 8-bit variable size **per registered event** and increase the access time (though I think we can disregard the access time for these)

No. You want to dump as much work on user code as possible.
You only need one list: of mcumgr callbacks from various subsystems of an app.
There should be Kconfig options to disable notify function from being invoked, in case your app is not interested at some point at all.
The app callback should be responsible for reacting to codes, even maybe having some state machine to take care of it.

For example app callback can have something like this:
```
callback(id, prev_callback_result, arg) {
    ...
    if (just_let_it_go) {
        return OK;
    }

    if (i_dont_like_fs_now && ((MCUMGR_MGMT_NOTIFY_AREA_GET(id) == MCUMGR_MGMT_NOTIFY_AREA_FS)) {
        return NO;
    }

    if (MCUMGR_MGMT_NOTIFY_AREA_GET(id) == MCUMGR_MGMT_NOTIFY_AREA_FS) {
        switch(MCUMGR_MGMT_NOTIFY_POINT_GET(id)) {
        case MCUMGR_MGMT_NOTIFY_FS_DOWNLOAD:
        {
            fs_notify_ctxt *fnc = arg;
            if (arg->token == current_token && is_download_path_allowed(current_token, fnc->path)) {
                if (offset == 0) {
                    global_state_busy = true;
                } else {
                    global_state_busy = false;
                }
                return OK;
            } else {
                return ACCESS_DENIED;
            }
        }
        default:
            ... //some def action for other points;
        } // end switch;
    if (MCUMGR_MGMT_NOTIFY_AREA_GET(id) == MCUMGR_MGMT_NOTIFY_AREA_OS && 
        MCUMGR_MGMT_NOTIFY_POINT_GET(id)  == MCUMGR_MGMT_NOTIFY_POINT_RESET_ASK) {
        if (global_state_busy == true) {
            return NO;
        }
    }
    return prev_callback_result;
 }
```

I am not even sure we need slist for that, because we may actually make callback registration function to return previous callback and make the user code to call the returned pointer. Little bit risky, but probably more efficient and flexible.

Reply:
This is at odds with the style of callbacks in zephyr https://docs.zephyrproject.org/latest/connectivity/networking/api/net_mgmt.html#listening-to-network-events
It does not make sense to invoke a callback (or multiple) every time a chunk of data is received in img_mgmt for a bunch of handlers that do not need it, that's 1024 pointless callbacks per handler for a 256KB image upload with 256 byte MTU.

Reply:
> This will also be called every time callback gets entered, so 1024 times:

The callback system will be called 1024 times, the correct handler will be called 1024 times. If there's 10 handlers with only 1 being needed for the check, the overhead of that without using a bitmask is 9 * 1024 = 9216. That overhead is huge.

> is up to callback

That is down to the implementation, for notifications for file access, if there are 10 handlers and one denies access to the file, there is no point in running the other handlers because one of those has already denied access to it, the callback is specifically for checking if file access is allowed or not, but there may be a callback that is used to inform of something, so even if a handler fails it will still call the others.

Reply:
> > This will also be called every time callback gets entered, so 1024 times:
> 
> The callback system will be called 1024 times, the correct handler will be called 1024 times. If there's 10 handlers with only 1 being needed for the check, the overhead of that without using a bitmask is 9 * 1024 = 9216. That overhead is huge.

You have missed this one (https://github.com/zephyrproject-rtos/zephyr/pull/50578#discussion_r987902648): 

Then we can have per-group/per-subsystem callback, but we should not block app from registering the same callback function in all of them and take the consequences, which means processing them in single state machine.

> That is down to the implementation, for notifications for file access, if there are 10 handlers and one denies access to the file, there is no point in running the other handlers because one of those has already denied access to it, the callback is specifically for checking if file access is allowed or not, but there may be a callback that is used to inform of something, so even if a handler fails it will still call the others. As a real world example you might have 2 handlers, one for access checking, one for rewriting the path, why would the path be rewritten if the access check fails? It is literally a complete waste of time, processing and power.

That is not your decision, that is not my decision, that is not decision MCUmgr should take. MCUmgr should get the code and behave as the code for specific point defines. Callbacks should only report something or return specific code according to logic within them. It is on the callback to change error level, if desired; callback may also decide whether it will call the next one, or allow the next one to be called, or just end the sequence and return code.

> As a real world example you might have 2 handlers, one for access checking, one for rewriting the path, why would the path be rewritten if the access check fails? It is literally a complete waste of time, processing and power.

If callback does it, then it does it and it is beyond us. In your real world example callback removing file may be called first, before the one that gives access; it may also, due to programmers error actually override the one that would give access with denial, because it failed to find the file. 
Neither of proposed solutions warranties order of callback execution. The one where callback decides, by calling next in order, does provide limited ability to callback to decide (at cost of stack), because callback may at any point call next in queue and decide what to do with returned value.

Reply:
Callback system updated to use zephyr callback design